File: others/Hash Tables/Sherlock and Anagrams/solve.swift
//Not optimal!

extension Character {
	var isAscii: Bool {
		return unicodeScalars.allSatisfy { $0.isASCII }
	}
	var ascii: UInt32? {
		return isAscii ? unicodeScalars.first?.value : nil
	}
}

func sherlockAndAnagrams(s: String) -> Int {
	let startChar = Unicode.Scalar("a").value
	let endChar = Unicode.Scalar("z").value
	let alphabet = (startChar...endChar).map { return String(Unicode.Scalar($0)!) }


	var signatures: [String: Int] = [:]
	var signature = Array(repeating: 0, count: alphabet.count)

	s.forEach { signature[Int($0.asciiValue! - Character(alphabet[0]).asciiValue!)] += 1  }

	for start in 0..<s.count {
		for finish in start..<s.count {
			var signature = Array(repeating: 0, count: alphabet.count)
			for index in start..<finish+1 {
				let letter = Array(s)[index]
				signature[Int(letter.ascii! - Character(alphabet[0]).ascii!)] += 1
			}
			let signatureKey = signature.map { return String($0) }.joined()
			let val = signatures[signatureKey] ?? 0
			signatures[signatureKey] = val + 1
		}
	}

	var result = 0
	for count in signatures.values {
		result += count*(count - 1) / 2
	}
	return result
}

File: others/Hash Tables/Sherlock and Anagrams/solve.py
import string

signatures = {}
ALPHABET = string.ascii_lowercase
signature = [0 for _ in ALPHABET]

for letter in s:
  signature[ord(letter)-ord(ALPHABET[0])] += 1

# iterate over all substrings of s
for start in range(len(s)):
  for finish in range(start, len(s)):
    # initialize substring signature
    signature = [0 for _ in ALPHABET]
    for letter in s[start:finish+1]:
      signature[ord(letter)-ord(ALPHABET[0])] += 1
    # tuples are hashable in contrast to lists
    signature = tuple(signature)
    print(signature)
    signatures[signature] = signatures.get(signature, 0) + 1


res = 0
for count in signatures.values():
  res += count*(count-1)/2

print(int(res))